token-swap/program/src/processor.rs

```diff
@@ -218,23 +227,26 @@ impl Processor {
         let destination_info = next_account_info(account_info_iter)?;
         let token_program_info = next_account_info(account_info_iter)?;
 
-        let token_swap = State::deserialize(&swap_info.data.borrow())?.token_swap()?;
+        let token_swap = SwapInfo::unpack(&swap_info.data.borrow())?;
+        if !token_swap.is_initialized {
+            return Err(SwapError::AlreadyInUse.into());
+        }
```

Comment:
We do this check several times; maybe time for an `unpack_checked` method? (Or an `unpack_unchecked` method for the init instruction, if you prefer)

Reply:
Also, should this be `SwapError::InvalidState`?

Reply:
I'll follow the `unpack_unchecked` model from `Token`, and absolutely, `InvalidState` is a better error